### PR TITLE
supersonic: add darwin support

### PR DESCRIPTION
**Pull request by vs49688**
## Description of changes

Build supersonic on Darwin.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vs49688** [comment]
@ofborg build supersonic supersonic-wayland

**Comment by vs49688** [comment]
This requires `UserNotifications`, which isn't in the x86_64-darwin SDK, but is in 11.0.

However, using 11.0 breaks the `pkgs/by-name` check (because it's `darwin.apple_sdk_11_0.callPackage` instead of just plain `callPackage`). Rock, meet hard place.

@infinisil sorry to bug you with this triviality, but what's the correct solution here?

**Comment by wegank** [comment]
> However, using 11.0 breaks the `pkgs/by-name` check (because it's `darwin.apple_sdk_11_0.callPackage` instead of just plain `callPackage`). Rock, meet hard place.

How?

**Comment by infinisil** [comment]
You can generally always rewrite it to work with the standard `callPackage`. In this case:

```nix
{
  lib,
  stdenv,
  buildGoModule,
  darwin,
  ...
}:
let
  inherit (darwin.apple_sdk_11_0.frameworks)
    UserNotifications;
in
buildGoModule {
  # ...
  buildInputs = [
    # ...
  ]
  ++ lib.optional stdenv.isDarwin UserNotifications;
}
```

File: pkgs/by-name/su/supersonic/package.nix [hunk 1]
```diff
@@ -35,9 +42,18 @@ buildGoModule rec {
   buildInputs = [
     libglvnd
     mpv
+  ] ++ lib.optionals stdenv.isLinux [
     xorg.libXxf86vm
     xorg.libX11
-  ] ++ (glfw3.override { inherit waylandSupport; }).buildInputs;
+  ]
+  ++ (builtins.filter (p: !(lib.hasPrefix "apple-framework" p.name))
+    (glfw3.override { inherit waylandSupport; }).buildInputs)
```
Comment (wegank) [inline pkgs/by-name/su/supersonic/package.nix:50]: I'd prefer the `buildInputs` to be explicit - the workaround there is terribly hacky.

Comment (vs49688) [reply pkgs/by-name/su/supersonic/package.nix]: Fair point, I'm not too fussed either way. I see you've already changed it 👍

**Comment by vs49688** [comment]
> > However, using 11.0 breaks the `pkgs/by-name` check (because it's `darwin.apple_sdk_11_0.callPackage` instead of just plain `callPackage`). Rock, meet hard place.
> 
> How?

I was apprehensive about passing in the `darwin` directly. That's not usually how I've usually seen it done. Also, using 
`darwin.apple_sdk.callPackage` always strikes me as kind of dodgy for a package that's not darwin-only, hence why I was wondering if there's a better way.

**Comment by wegank** [comment]
> Also, using `darwin.apple_sdk.callPackage` always strikes me as kind of dodgy for a package that's not darwin-only, hence why I was wondering if there's a better way.

I would expect `stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv` to work for most of the time. Sometimes this would be more difficult, as a lot of packages get overridden in `pkgs/os-specific/darwin/apple-sdk-11.0/default.nix`.

**Pull request merged**